---
title: "Riboseq_quantseq_figures"
author: "Oscar Wilkins"
date: "19/03/2021"
output: html_document
---

```{r setup, include=FALSE, message=FALSE, warning=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = FALSE, warning=FALSE)
if (!require("pacman")) install.packages("pacman")
library(pacman)
p_load("tidyverse", "here")
```

## 6 Way Heatmap of differential RFP data

This code generates a 6-way heatmap of the RFP data. It reads in the processed data, which is produced by the script ""all_riboseq_quantseq_preprocessing.R" 

```{r 6wayRFP}

sq_lg2 <- read_csv(paste0(here(), "/data/square_lg2.csv"))

all_comparisons <- sq_lg2 %>%
  pivot_longer(cols = c("A", "B", "C", "D", "E", "G"), 
               names_to="reference",
               values_to = "log2fold") %>%
  filter(!is.na(log2fold))

ggplot(all_comparisons, aes(x = reference, y = comparison)) + 
  geom_tile(aes(fill=log2fold)) +
  scale_fill_gradient2(low="blue", mid= "white" ,high="red", midpoint=0,
                       limits = c(-2,2), breaks =c(-2, -1, 0, 1, 2))  +
  theme_minimal() +
  ggtitle("Pair-wise comparison of number of RFPs aligning to MAPT gene") +
  xlab("Reference Sample") +
  ylab("Comparison Sample")

```

## Riboseq volcano plot

This reads in the DESEQ2 values from data generated by "all_riboseq_quantseq_preprocessing.R". Produces a volcano plot, and states MAPT's rank.

```{r rfp_volcano}

rfp_data <- read_csv(paste0(here(), "/data/RFP_B_vs_A_volcano_data.csv"))

min_counts_per_sample_average <- 10
min_counts <- min_counts_per_sample_average*17

ggplot(rfp_data, aes(log2FoldChange, minus_log10_p, colour = color, name=name)) +
  geom_point(size=1, alpha = 1) +
  ggtitle(paste0("Volcano plot of condition B versus A with minimum \n", min_counts, " counts total across 17 samples")) +
  scale_colour_identity() +
  theme_minimal() +
  geom_text(aes(label=name),hjust=0.6, vjust=1.2, size =3)

# State rank of mapt

rfp_data2 <- rfp_data %>%
  arrange(log2FoldChange) %>%
  mutate(rank = 1:n())

mapt_rank = rfp_data2$rank[which(rfp_data2$gene_name == "MAPT")]

print(paste0("MAPT ranked ", mapt_rank, " of ", nrow(rfp_data2)))

```


## Quantseq volcano plot

This reads in DESEQ2 values from data generated by "quantseq_preprocessing.R"

```{r quantseq}

quantseq_data <- read_csv(paste0(here(), "/data/quantseq_B_vs_A_volcano_data.csv"))

min_counts_quantseq <- 1000

ggplot(quantseq_data, aes(log2FoldChange, minus_log10_p, color = colour2, alpha =1, name = name)) +
  geom_point() +
  scale_color_identity() +
  scale_alpha_identity() +
  geom_text(aes(label=name),hjust=0.6, vjust=1.2, size =3) +
  ggtitle(paste0("Quantseq B versus A for min total \ncounts of ", min_counts_quantseq, " across 18 samples")) +
  theme_minimal()

```


## Riboseq quality control

```{r}

min_length <- 29
max_length <- 35

info <- read_tsv(paste0(here(), "/data/longest_proteincoding_transcript_hs_details.txt")) # file containing all 

file <- paste0(here(), "/data/A1.Aligned.toTranscriptome.out.bam.bed.gz")

data <- read_tsv(file, col_names = c("transcript_id", "zero_based_start", "end", 
                                     "name", "score", "strand")) %>%
  dplyr::filter(strand == "+") %>%
  mutate(start = zero_based_start + 1) %>%
  mutate(length = end-start+1) %>%
  inner_join(info, by = "transcript_id") %>%
  mutate(frame = (start-cds_start)%%3) %>% # 5' frame
  group_by(frame, length) %>%
  mutate(number = n()) %>%
  ungroup() %>%
  dplyr::filter(length >= min_length & length <= max_length) %>%
  dplyr::select(length, frame, number) %>%
  unique()

ggplot(data, aes(x = length, y = number, fill = factor(frame))) +
  geom_bar(position="dodge", stat="identity") +
  xlab("Read length") +
  theme(axis.text.y=element_blank(),
        axis.title.y=element_blank(),
        axis.ticks.y=element_blank()) +
  ggtitle("Relative number of reads with 5' end in each frame")


# And metaplot

meta <- rep(0,200)

bed <- read_tsv(paste0(here(), "/data/G2.Aligned.toTranscriptome.out.bam.bed.gz_forParaclu.bed.gz"), 
                col_names = c("transcript_id", "start", "end", "gene_id", "counts", "strand")) %>%
    left_join(info, by = c("transcript_id", "gene_id")) %>%
    group_by(transcript_id) %>%
    mutate(sum_tx = sum(counts)) %>%
    ungroup() %>%
    mutate(normalised_counts = counts/sum_tx)

for(row in 1:nrow(bed)){
    this_cds_start <- bed$cds_start[row]
    this_cds_end <- bed$cds_end[row]
    this_pos <- bed$start[row]
    this_counts <- bed$normalised_counts[row]
    this_cds_length <- bed$cds_length[row]
    this_tx_length <- bed$tx_length[row]
    
    # 50 buckets for 5 and 3' utr, 100 for cds
    
    # if in 5' utr
    if(this_pos < this_cds_start){
      metapos <- ceiling(49*this_pos/this_cds_start)
      meta[metapos] <- meta[metapos] + this_counts
      
    }else if(this_pos >= this_cds_start & this_pos <= this_cds_end){
      metapos <- 50+floor(100*(this_pos-this_cds_start)/this_cds_length)
      meta[metapos] <- meta[metapos] + this_counts
      
    }else if(this_pos > this_cds_end){
      metapos <- 150 + floor(50*(this_pos-this_cds_end)/(this_tx_length-this_cds_end))
      meta[metapos] <- meta[metapos] + this_counts
    }
}
  
meta_data <- data.frame(Counts = meta) %>%
  rownames_to_column(var = "Positionwrong") %>%
  mutate(Position = as.numeric(Positionwrong))

ggplot(meta_data, aes(x = Position, y = Counts))+
  geom_area() +
  ggtitle("Distribution of counts")


## extra metaplots not in paper

data <- read_tsv(file, col_names = c("transcript_id", "zero_based_start", "end", 
                                     "name", "score", "strand")) %>%
  dplyr::filter(strand == "+") %>%
  mutate(start = zero_based_start + 1) %>%
  mutate(length = end-start+1) %>%
  inner_join(info, by = "transcript_id") 

df_start <- data %>%
  mutate(from_start = start-cds_start) %>%
  group_by(from_start) %>%
  mutate(n = n()) %>%
  select(from_start, n) %>%
  unique() 

ggplot(df_start %>% filter(abs(from_start) < 100), aes(x = from_start, y = n)) +
  geom_bar(stat="identity") +
  ggtitle("Metaplots around start for A1 - all read lengths so strong periodicity NOT expected")

df_end <- data %>%
  mutate(from_end = start-cds_end) %>%
  group_by(from_end) %>%
  mutate(n = n()) %>%
  select(from_end, n) %>%
  unique() 

ggplot(df_end %>% filter(abs(from_end) < 100), aes(x = from_end, y = n)) +
  geom_bar(stat="identity")+
  ggtitle("Metaplots around end for A1 - all read lengths so strong periodicity NOT expected")


```
